src/algorithms/calorimetry/ImagingTopoCluster.h

```diff
@@ -198,35 +239,57 @@ namespace eicrecon {
     }
 
     // grouping function with Breadth-First Search
-    void bfs_group(const edm4eic::CalorimeterHitCollection &hits, std::set<std::size_t> &group, std::size_t idx, std::vector<bool> &visits) const {
-      visits[idx] = true;
+    template<typename Compare>
+    void bfs_group(const edm4eic::CalorimeterHitCollection &hits, std::set<std::size_t,Compare>& indices, std::vector<std::size_t> &group, const std::size_t idx) const {
 
       // not a qualified hit to participate clustering, stop here
       if (hits[idx].getEnergy() < m_cfg.minClusterHitEdep) {
         return;
       }
 
-      group.insert(idx);
-      size_t prev_size = 0;
-
+      std::size_t prev_size = 0;
       while (prev_size != group.size()) {
-        prev_size = group.size();
-        for (std::size_t idx1 : group) {
-          // check neighbours
-          for (std::size_t idx2 = 0; idx2 < hits.size(); ++idx2) {
-            // not a qualified hit to participate clustering, skip
-            if (hits[idx2].getEnergy() < m_cfg.minClusterHitEdep) {
+        auto prev_group = group;
```

Comment:
I wonder now if we need a copy of the group. As we add hits, group will expand but idx1 is always before the insertion point and remains valid. End is always invalidated but we don't store it. At the end of each loop iteration the != end() condition can be safely evaluated. Without prev_group we'd have fewer copy operations and fewer while loop iterations, maybe only one though I can prove it right now.

Reply:
Implemented.